src/test/python/test_signer.py

```diff
@@ -0,0 +1,166 @@
+        signer.__get_hashed_payload(b'{""}')
+        assert excepinfo.type is json.JSONDecodeError
+
+
+def test_none_payload():
+    assert signer.__get_hashed_payload(b'{}') == \
+           signer.__get_hashed_payload(None)
+
+
+def test_unordered_payload():
+    assert signer.__get_hashed_payload(b'{"name": "rattan", "a": "value"}') == \
+           signer.__get_hashed_payload(b'{"a": "value", "name": "rattan"}')
+
+
+def test_nested_payload():
+    input_json = \
+        b"""
+        {
+            "name": "rattan",
+            "a": "value",
+            "c": {
+                "za": "value",
+                "a": "hello"
+            }
+        }
+        """
+
+    sorted_json = \
+        b"""
+        {
+            "a": "value",
+            "c": {
+                "a": "hello",
+                "za": "value"
+            },
+            "name": "rattan"
+        }
+        """
+
+    assert signer.__get_hashed_payload(input_json) == \
+           signer.__get_hashed_payload(sorted_json)
+
+
+@pytest.mark.parametrize('url, expected_encoded_url', [
+    ('/identity blah/', '/identity%20blah/'),
+    ('/identity-blah/', '/identity-blah/'),
+    ('/', '/'),
+    ('/identityABC/', '/identityABC/'),
+])
+def test_uri_endoing(url, expected_encoded_url):
+    assert signer.__encode_uri(url) == expected_encoded_url
+
+
+@freeze_time()
+def test_signature_material():
+    method = "POST"
+    url = "https://delta.covata.io/v1/secrets?hello=w+orld"
+    headers = {
+        "Content-Type": "application/json"
+    }
+    payload = \
+        b"""
+        {
+            "name": "rattan",
+            "a": "value",
+            "c": {
+                "za": "value",
+                "a": "hello"
+            }
+        }
+        """
+
+    url_parsed = urllib.parse.urlparse(url)
+    cvt_date = datetime.utcnow().strftime(CVT_DATE_FORMAT)
+    headers_ = dict(headers)
+    headers_["Cvt-Date"] = cvt_date
+    uri = signer.__encode_uri("/".join(url_parsed.path.split("/")[2:]))
+    query = url_parsed.query.replace("+", "%20")
+    canonical_headers = 'content-type:application/json\n cvt-date:' \
+                        + cvt_date
+    signed_headers = 'content-type;cvt-date'
+    hashed_payload = \
+        '758ffa295b9a475f04aef51abd60563dabb8df1988cf6d62b9298b1d5ba6b8bf'
+    expected_content_type = 'application/json'
+    materials_from_signer = \
+        signer.__get_signature_materials(method, url, headers, payload)
+
+    assert(materials_from_signer.method is method)
+    assert(materials_from_signer.uri == uri)
+    assert (materials_from_signer.headers_ == headers_)
+    assert (materials_from_signer.query_params == query)
+    assert (materials_from_signer.canonical_headers == canonical_headers)
+    assert (materials_from_signer.signed_headers == signed_headers)
+    assert (materials_from_signer.hashed_payload == hashed_payload)
+    assert(materials_from_signer.headers_['Content-Type'] is expected_content_type)
```

Comment:
On second thoughts do we need it? since we are already doing the assertions on the headers which covers content type and cvt date. I removed the explicit assertion for content type